homeassistant/components/camera/media_source.py

```diff
@@ -0,0 +1,103 @@
+"""Expose cameras as media sources."""
+from __future__ import annotations
+
+from typing import Optional, cast
+
+from homeassistant.components.media_player.const import (
+    MEDIA_CLASS_APP,
+    MEDIA_CLASS_VIDEO,
+)
+from homeassistant.components.media_player.errors import BrowseError
+from homeassistant.components.media_source.error import Unresolvable
+from homeassistant.components.media_source.models import (
+    BrowseMediaSource,
+    MediaSource,
+    MediaSourceItem,
+    PlayMedia,
+)
+from homeassistant.components.stream.const import FORMAT_CONTENT_TYPE, HLS_PROVIDER
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_component import EntityComponent
+
+from . import Camera, _async_stream_endpoint_url
+from .const import DOMAIN, STREAM_TYPE_HLS
+
+
+async def async_get_media_source(hass: HomeAssistant) -> CameraMediaSource:
+    """Set up camera media source."""
+    return CameraMediaSource(hass)
+
+
+class CameraMediaSource(MediaSource):
+    """Provide camera feeds as media sources."""
+
+    name: str = "Camera"
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize CameraMediaSource."""
+        super().__init__(DOMAIN)
+        self.hass = hass
+
+    async def async_resolve_media(self, item: MediaSourceItem) -> PlayMedia:
+        """Resolve media to a url."""
+        component: EntityComponent = self.hass.data[DOMAIN]
+        camera = cast(Optional[Camera], component.get_entity(item.identifier))
+
+        if not camera:
+            raise Unresolvable(f"Could not resolve media item: {item.identifier}")
+
+        if camera.frontend_stream_type != STREAM_TYPE_HLS:
+            raise Unresolvable("Camera does not support HLS streaming.")
+
+        try:
+            url = await _async_stream_endpoint_url(self.hass, camera, HLS_PROVIDER)
+        except HomeAssistantError as err:
+            raise Unresolvable(str(err)) from err
+
+        return PlayMedia(url, FORMAT_CONTENT_TYPE[HLS_PROVIDER])
+
+    async def async_browse_media(
+        self,
+        item: MediaSourceItem,
+    ) -> BrowseMediaSource:
+        """Return media."""
+        if item.identifier:
+            raise BrowseError("Unknown item")
+
+        if "stream" not in self.hass.config.components:
+            raise BrowseError("Stream integration is not loaded")
+
+        # Root. List cameras.
+        component: EntityComponent = self.hass.data[DOMAIN]
+        children = []
+        for camera in component.entities:
+            camera = cast(Camera, camera)
+
+            if camera.frontend_stream_type != STREAM_TYPE_HLS:
```

Comment:
Is making this work just a matter of frontend support or are there other fundamental issues you forsee here?

Reply:
It should be able to work with other types but right now I mimicked what the `camera.play_stream` service is doing.